cli/azd/pkg/input/console.go

```diff
@@ -63,13 +68,35 @@ func (c *AskerConsole) SetWriter(writer io.Writer) {
 // Prints out a message to the underlying console write
 func (c *AskerConsole) Message(ctx context.Context, message string) {
 	// Disable output when formatting is enabled
-	if c.formatter == nil || c.formatter.Kind() == output.NoneFormat {
+	if c.formatter != nil && c.formatter.Kind() == output.JsonFormat {
+		// we do not expect formatter.Format to fail on valid input. Discard the error to appease the linter (note
+		// that we implicitly discard the error from fmt.Fprintln later, as well).
+		_ = c.formatter.Format(c.eventForMessage(message), c.writer, nil)
```

Comment:
Continuing [conversation](https://github.com/Azure/azure-dev/pull/925#discussion_r1000872506) about whether this line should be minified:

> I don't think so. I think any consumer is just going to need to read an object at a time from `stdout` and the intermixed newlines won't cause an issue. FWIW, in an earlier iteration I did just write this as a single line (eschewing the formatter API and just calling `json.Marshal` myself so there was no indentation) but I realized that would be different from the final result for most of our commands (where we do format the object across multiple lines and indented).
> 
> This also makes reading the JSON output as a human so much easier, which I think is a nice plus.

Can we guarantee that there isn't any type of buffering between us and the actual consumer? Unless we're sending discrete events, it could happen that the consumer doesn't see each JSON delivered atomically. It also makes it not possible to reparse an `azd` stderr output JSON at a later stage if the results were to be piped into a file.

I think guaranteeing newline separation makes the contract machine-readable. Human readability does suffer but the idea of json output is to be consumed by machines.

Reply:
> Can we guarantee that there isn't any type of buffering between us and the actual consumer? Unless we're sending discrete events, it could happen that the consumer doesn't see each JSON delivered atomically. It also makes it not possible to reparse an azd stderr output JSON at a later stage if the results were to be piped into a file.

I think I'm confused about what you mean here by buffering and how newlines would play into it.  Today we do send discrete events (every call to `.Message()` yeilds a single call to `formatter.Format` which writes a complete object to stderr.  If stderr is buffered, the producer (azd) may only write a partial message and then block while it waits for the consumer to read the data out of the buffer, but that blocking could happen anywhere in the message, correct?

I agree it would be bad if two calls (in different goroutines) are racing and both try to write a message at the same time, but I'm not sure saying that every message will be a single line actually helps us there?  Couldn't messages still be interleaved in that case, even if we didn't add newlines?

I also don't understand how it makes it not possible to reparse the output if it was written to a file.  I assume you are worried about something like this:

```
$ azd infra create --output=json 2>events.txt 1>nul
```

And then trying to read events.txt after the fact.  But it feels to me that you can parse this in the same way you would parse the stream normally, just read a JSON object at a time from the byte stream, until you reach EOF?

> I think guaranteeing newline separation makes the contract machine-readable. Human readability does suffer but the idea of json output is to be consumed by machines.

Would you also not render newlines for the value written to `stdout`?  This is different than what `az` does (and most tools I've used)?  I guess it's not the end of the world because you can always get back to a human readable form using something like `| jq .` but it still feels a little weird to me.

Reply:
Wei and I chatted about this on a Teams call.  His concerns were twofold:

1. Looking across the ecosystem, it seemed like most tools would write these diagnostics messages as a single line and it seemed weird for us to divert.
2. In cases where you may have a JSON reader that works on strings, not streams, it may be harder to read set of objects from a string, since these libraries might only expect a single object (since they take a string).  If each object was on its own line, you could split the string on newlines to get an array of strings, each which contains a single object, and that would be easier to parse with these libraries.

Given that these events are targeted at machines, not humans, we decided it would be OK to put them on a single line instead of formatting them nicely.

We'll continue to format the final result with newlines and indentation, since that's much nicer UX for folks and also we don't run into the above problem since there would only ever be a single object written to stdout.

Reply:
Addressed with [1b2a02a](https://github.com/Azure/azure-dev/pull/925/commits/1b2a02ab327e268c8e5a1fd3f238fe503da4acc5).